src/coreclr/tools/Common/TypeSystem/Common/Utilities/GCPointerMap.Algorithm.cs

```diff
@@ -115,22 +111,47 @@ public static GCPointerMap FromThreadStaticLayout(MetadataType type)
                 TypeDesc fieldType = field.FieldType;
                 if (fieldType.IsGCPointer)
                 {
-                    builder.MarkGCPointer(field.Offset.AsInt);
+                    builder.MarkGCPointer(field.Offset.AsInt + baseOffset);
                 }
                 else if (fieldType.IsValueType)
                 {
                     var fieldDefType = (MetadataType)fieldType;
                     if (fieldDefType.ContainsGCPointers)
                     {
                         GCPointerMapBuilder innerBuilder =
-                            builder.GetInnerBuilder(field.Offset.AsInt, fieldDefType.InstanceByteCount.AsInt);
+                            builder.GetInnerBuilder(field.Offset.AsInt + baseOffset, fieldDefType.InstanceByteCount.AsInt);
                         FromInstanceLayoutHelper(ref innerBuilder, fieldDefType);
                     }
                 }
             }
+        }
+
+        /// <summary>
+        /// Computes the GC pointer map of the thread static region of the type.
+        /// </summary>
+        public static GCPointerMap FromThreadStaticLayout(MetadataType type)
+        {
+            GCPointerMapBuilder builder = new GCPointerMapBuilder(type.ThreadGcStaticFieldSize.AsInt, type.Context.Target.PointerSize);
+
+            MapThreadStaticsForType(builder, type, baseOffset: 0);
 
             Debug.Assert(builder.ToGCMap().Size * type.Context.Target.PointerSize >= type.ThreadGcStaticFieldSize.AsInt);
             return builder.ToGCMap();
         }
+
+        public static GCPointerMap FromInlinedThreadStatics(
+            List<MetadataType> types,
+            Dictionary<MetadataType, int> offsets,
+            int threadStaticSize,
+            int pointerSize)
+        {
+            GCPointerMapBuilder builder = new GCPointerMapBuilder(threadStaticSize, pointerSize);
+            foreach (var type in types)
+            {
+                MapThreadStaticsForType(builder, type, offsets[type]);
```

Comment:
`builder` is a valuetype. This code doesn't introduce a GC hole only by accident because we don't try to optimize out the array allocation in the builder for small maps.

Reply:
The builder seems to be an immutable struct that wraps an array. I assumed that it is the intent. Such patterns happen.

I'll change it to be passed by reference. Thanks!